### ISSUE TITLE
NPE when running test suite on Clojure 1.8+, probably related to #63

### ISSUE THREAD

**Issue by aviflax** [weavejester/eftest#65]
I first saw this when I ran the test suite with Clojure 1.10 on a lark, when I was working on something else. But then, when I upgraded my project to the latest release (0.5.6, incorporating #63) I started seeing it as part of every run of my project’s test suite:

```
$ clojure -A:test:test/run

53/53   100% [==================================================]  ETA: 00:00
Exception in thread "main" java.lang.NullPointerException
	at clojure.lang.Numbers.ops(Numbers.java:1068)
	at clojure.lang.Numbers.add(Numbers.java:153)
	at eftest.report.pretty$eval2315$fn__2317.invoke(pretty.clj:150)
	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
	at eftest.report.progress$eval2403$fn__2404.invoke(progress.clj:65)
	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
	at fc4.test_runner.runner$multi_report$fn__3069.invoke(runner.clj:23)
	at clojure.test$do_report.invokeStatic(test.clj:357)
	at clojure.test$do_report.invoke(test.clj:351)
	at eftest.runner$run_tests.invokeStatic(runner.clj:202)
	at eftest.runner$run_tests.invoke(runner.clj:166)
	at fc4.test_runner.runner$run_tests.invokeStatic(runner.clj:53)
	at fc4.test_runner.runner$run_tests.invoke(runner.clj:51)
	at fc4.test_runner.runner$_main.invokeStatic(runner.clj:56)
	at fc4.test_runner.runner$_main.invoke(runner.clj:55)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.lang.Var.applyTo(Var.java:705)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.main$main_opt.invokeStatic(main.clj:491)
	at clojure.main$main_opt.invoke(main.clj:487)
	at clojure.main$main.invokeStatic(main.clj:598)
	at clojure.main$main.doInvoke(main.clj:561)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:705)
	at clojure.main.main(main.java:37)
```

I wanted to debug this a little, maybe even figure out how to fix it. So I added some Leiningen profiles to my local checkout of eftest, one for each of Clojure 1.8, 1.9, and 1.10. When I run `lein test` and the tests are run on Clojure 1.7, everything looks fine. But when I run `lein with-profile 1.8 test` (or 1.9 or 1.10) I get the same exception, with a longer stacktrace:

<details>
<summary>Longer stacktrace</summary>

```
...
lein test :only eftest.runner-test/test-fail-multi

ERROR in (test-fail-multi) (Numbers.java:1068)
Uncaught exception, not in assertion.
expected: nil
actual: java.lang.NullPointerException: null
 at clojure.lang.Numbers.ops (Numbers.java:1068)
    clojure.lang.Numbers.add (Numbers.java:153)
    eftest.report.pretty$eval3044$fn__3046.invoke (pretty.clj:151)
    clojure.lang.MultiFn.invoke (MultiFn.java:229)
    eftest.report.progress$eval3132$fn__3133.invoke (progress.clj:65)
    clojure.lang.MultiFn.invoke (MultiFn.java:229)
    clojure.test$do_report.invokeStatic (test.clj:357)
    clojure.test$do_report.invoke (test.clj:351)
    eftest.runner$run_tests.invokeStatic (runner.clj:202)
    eftest.runner$run_tests.invoke (runner.clj:166)
    eftest.runner_test$test_run_tests$fn__3349.invoke (runner_test.clj:36)
    eftest.runner_test$with_test_out_str_STAR_$fn__3341.invoke (runner_test.clj:22)
    eftest.runner_test$with_test_out_str_STAR_.invokeStatic (runner_test.clj:21)
    eftest.runner_test$with_test_out_str_STAR_.invoke (runner_test.clj:19)
    eftest.runner_test$test_run_tests.invokeStatic (runner_test.clj:36)
    eftest.runner_test$test_run_tests.invoke (runner_test.clj:30)
    eftest.runner_test$test_run_tests.invokeStatic (runner_test.clj:32)
    eftest.runner_test$test_run_tests.invoke (runner_test.clj:30)
    eftest.runner_test$fn__3366.invokeStatic (runner_test.clj:55)
    eftest.runner_test/fn (runner_test.clj:53)
    clojure.test$test_var$fn__9707.invoke (test.clj:717)
    clojure.test$test_var.invokeStatic (test.clj:717)
    clojure.test$test_var.invoke (test.clj:708)
    clojure.test$test_vars$fn__9733$fn__9738.invoke (test.clj:735)
    clojure.test$default_fixture.invokeStatic (test.clj:687)
    clojure.test$default_fixture.invoke (test.clj:683)
    clojure.test$test_vars$fn__9733.invoke (test.clj:735)
    clojure.test$default_fixture.invokeStatic (test.clj:687)
    clojure.test$default_fixture.invoke (test.clj:683)
    clojure.test$test_vars.invokeStatic (test.clj:731)
    clojure.test$test_all_vars.invokeStatic (test.clj:737)
    clojure.test$test_ns.invokeStatic (test.clj:758)
    clojure.test$test_ns.invoke (test.clj:743)
    user$eval224$fn__285.invoke (form-init4797442981940918597.clj:1)
    clojure.lang.AFn.applyToHelper (AFn.java:156)
    clojure.lang.AFn.applyTo (AFn.java:144)
    clojure.core$apply.invokeStatic (core.clj:667)
    clojure.core$apply.invoke (core.clj:660)
    leiningen.core.injected$compose_hooks$fn__154.doInvoke (form-init4797442981940918597.clj:1)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:665)
    clojure.core$apply.invoke (core.clj:660)
    leiningen.core.injected$run_hooks.invokeStatic (form-init4797442981940918597.clj:1)
    leiningen.core.injected$run_hooks.invoke (form-init4797442981940918597.clj:1)
    leiningen.core.injected$prepare_for_hooks$fn__159$fn__160.doInvoke (form-init4797442981940918597.clj:1)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.lang.AFunction$1.doInvoke (AFunction.java:31)
    clojure.lang.RestFn.invoke (RestFn.java:408)
    clojure.core$map$fn__5851.invoke (core.clj:2755)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.Cons.next (Cons.java:39)
    clojure.lang.RT.boundedLength (RT.java:1788)
    clojure.lang.RestFn.applyTo (RestFn.java:130)
    clojure.core$apply.invokeStatic (core.clj:667)
    clojure.test$run_tests.invokeStatic (test.clj:768)
    clojure.test$run_tests.doInvoke (test.clj:768)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:665)
    clojure.core$apply.invoke (core.clj:660)
    user$eval224$fn__297$fn__330.invoke (form-init4797442981940918597.clj:1)
    user$eval224$fn__297$fn__298.invoke (form-init4797442981940918597.clj:1)
    user$eval224$fn__297.invoke (form-init4797442981940918597.clj:1)
    user$eval224.invokeStatic (form-init4797442981940918597.clj:1)
    user$eval224.invoke (form-init4797442981940918597.clj:1)
    clojure.lang.Compiler.eval (Compiler.java:7176)
    clojure.lang.Compiler.eval (Compiler.java:7166)
    clojure.lang.Compiler.load (Compiler.java:7635)
    clojure.lang.Compiler.loadFile (Compiler.java:7573)
    clojure.main$load_script.invokeStatic (main.clj:452)
    clojure.main$init_opt.invokeStatic (main.clj:454)
    clojure.main$init_opt.invoke (main.clj:454)
    clojure.main$initialize.invokeStatic (main.clj:485)
    clojure.main$null_opt.invokeStatic (main.clj:519)
    clojure.main$null_opt.invoke (main.clj:516)
    clojure.main$main.invokeStatic (main.clj:598)
    clojure.main$main.doInvoke (main.clj:561)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.lang.Var.applyTo (Var.java:705)
    clojure.main.main (main.java:37)
```

</details>
&nbsp;

So I checked out line 150 of `pretty.clj` and saw this:

```clojure
149 (defmethod report :summary [{:keys [test pass fail error duration]}]
150   (let [total (+ pass fail error)
```

I figured that one of `pass | fail | error` was not in the input map.

So I mangled the method so I could do some println debugging, and I saw that indeed, when the NPE is thrown, the value of the input map is something like `{:type :summary, :duration 8.326198}` — so it’s missing all three keys (`:pass`, `:fail`, and `:error`).

That’s as far as I got. I have no ideas as to why the input map would sometimes be missing those keys, and why that would appear to occur on Clojure 1.8+ but not 1.7.

Any ideas? I’d be happy to try to contribute a fix, if I could get some help with the debugging.

**Comment by weavejester** [comment]
Thanks for reporting this. It looks like some changes have been made to Clojure's type hinting mechanism that stops an executor seeing a function as a `Callable` even when its type hinted:

```clojure
eftest.runner=> @(.submit e (reify Callable (call [_] (+ 1 1))))
2
eftest.runner=> @(.submit e ^Callable #(+ 1 1))
nil
```

Instead, it's seeing the function as a `Runnable`, and ignoring the return value.

This is particularly strange, because we can see a Callable type hint used in the Clojure source itself for this very purpose: https://github.com/clojure/clojure/blob/master/src/clj/clojure/core.clj#L6973

Why it works in core and not elsewhere I'm uncertain. Perhaps something to do with direct linking, which was added to `clojure.core` in 1.8.0?

In any case, I've got a fix by using `reify` to directly implement `Callable` and only `Callable`. I've added the fix to the master branch. Could you check to see if the fix works for you? If it does, I'll release 0.5.7.

**Comment by aviflax** [comment]
My pleasure. The fix works great! Thanks!

**Comment by aviflax** [comment]
BTW, apropos to this sort of thing, would you be interested in a contribution of a CircleCI config file that would run the test suite against multiple Clojure versions? (CircleCI offers gratis plans for open source projects.)

**Comment by weavejester** [comment]
I tend to prefer Travis for my open source projects. The default setup is more suitable for my purposes, which is useful when you have a large number of distinct projects.

For production projects, or weird builds, I use CircleCI instead.

**Comment by aviflax** [comment]
OK, I’ll take a look at Travis.

Looking forward to 0.5.7! 😉

**Comment by weavejester** [comment]
I've added some Travis tests and released 0.5.7.

**Comment by aviflax** [comment]
Excellent!
